clippy_lints/src/mut_key.rs

```diff
@@ -103,30 +103,63 @@ fn check_sig<'tcx>(cx: &LateContext<'tcx>, item_hir_id: hir::HirId, decl: &hir::
 fn check_ty<'tcx>(cx: &LateContext<'tcx>, span: Span, ty: Ty<'tcx>) {
     let ty = ty.peel_refs();
     if let Adt(def, substs) = ty.kind() {
-        if [sym::hashmap_type, sym::BTreeMap, sym::hashset_type, sym::BTreeMap]
+        let is_keyed_type = [sym::hashmap_type, sym::BTreeMap, sym::hashset_type, sym::BTreeSet]
             .iter()
-            .any(|diag_item| cx.tcx.is_diagnostic_item(*diag_item, def.did))
-            && is_mutable_type(cx, substs.type_at(0), span)
-        {
+            .any(|diag_item| cx.tcx.is_diagnostic_item(*diag_item, def.did));
+        if is_keyed_type && is_mutable_type(cx, substs.type_at(0), span) {
             span_lint(cx, MUTABLE_KEY_TYPE, span, "mutable key type");
         }
     }
 }
 
 fn is_mutable_type<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>, span: Span) -> bool {
     match *ty.kind() {
-        RawPtr(TypeAndMut { ty: inner_ty, mutbl }) | Ref(_, inner_ty, mutbl) => {
-            mutbl == hir::Mutability::Mut || is_mutable_type(cx, inner_ty, span)
-        },
```

Comment:
Can we name and document the function `is_mutable_type` more precisely? Then everything else here is more or less self-explanatory. Comments in the function should ideally be written with respect to what this function does, not the lint.

```rust
/// Returns true for types where the hash of an instance of the type can be modified using
/// interior mutability (without a mutable reference).
/// For example, this includes `Cell` and `Vec<Cell<_>>`.
fn type_has_interior_mutable_hash(...) {
```

Reply:
I think `is_interior_mutable_type` would be good - the lint cares about both `Hash` _and_ `Ord`, and I don't feel like bloating the function name with that much detail.  I've added a doc comment which hopefully explains things.  I'll leave this conversation unresolved in case you have more feedback.

Reply:
That's fine since it's isolated to this lint.